src/ceph-volume/ceph_volume/util/device.py

```diff
@@ -75,12 +80,25 @@ class Device(object):
         'vendor',
     ]
 
-    def __init__(self, path):
+    def __init__(self, path, lvs=None):
+        global global_lvs
+
         self.path = path
+
+        if lvm.is_pv(path):
+            self.lvs = lvs or lvm.get_lv(pv_name=path, lvs=global_lvs)
+        elif lvm.is_vg(os.path.basename(path)):
+            self.lvs = lvs or lvm.get_lv(vg_name=os.path.basename(path),
+                                         lvs=global_lvs)
+        elif lvm.is_lv(lv_name=path):
```

Comment:
Hmm this will list all PVs, all VGs and all LVs for each device. Not really the direction we want to go with this.

Reply:
True. I'm not sure if there is a way without it because I ended up writing this in my effort to get the PR past CI failures. I'll try to look if there's an alternative but I am anticipating that memoization will probably be the only way to optimize it.

This is one reason why in my head I keep thinking if having a global list would be nicer.

Reply:
Globals, memoizing (caching), are *really* hard to do them right. Where are we in the performance measurements? Is this 50% better already? What are we trying to solve?

Reply:
What I want is to avoid scanning and getting a list from scratch for PVs, VGs and LVs which can happen several times with in a command.

If we have these 3 tuples globally (or globally somwhere in `ceph_volume/lvm`, we can avoid totally avoid creating list from scratch.

Reply:
The original intend of this was to reduce the number of calls to lvm (see https://tracker.ceph.com/issues/37492). The hope was that this will also improve behaviour with damaged LVs.

I think however that the current `get_lv` and `lvs.get` is too complex to factor effectively. I'd think we should hand off the filtering on LVs to LVM itself with a handful of purpose build calls like [this one](https://github.com/ceph/ceph/pull/31410/files#diff-0d937eeea907d69c580b4c2fe1fdf096R1245). Not only will this improve maintainability, but will also simplify testing.

Reply:
Right now, I've made these lists global only within `utill/device.py`. I added commit just few hours ago; I'm testing it right now.

Reply:
@jan--f 

You are right about complexity of `get_lv`.

We can break `get_lv` and `lvs.get` to few methods like [`get_lv_by_name`](https://github.com/ceph/ceph/pull/31410/files#diff-0d937eeea907d69c580b4c2fe1fdf096R1245) and use the list memoized globally or within `api/lvm.py` instead. This way we would have best of both worlds (maintainability vs performance).